### PR TITLE
Batch edit for collection

### PR DESCRIPTION
**Pull request by eddierubeiz**
Make collection batch-editable.
Ref #964

### CODE REVIEW

**Comment by jrochkind** [comment]
The implementation here seems to me to be more complicated than it needs to be, but I want to spend some time with it. 

First though, a question about the intended spec -- is this intending to *replace* any existing collection ids with the new one(s) specified, or append the new one to the existing ones?

What does it do with existing multi-valued fields, as far as replacing existing values or appending to them?  This should probably do the same, and our test should probably ensure that it does that.  (Unless we have reason to want it to do someting different here? Check with Annabel?)

 So that's something that could be enhanced right now, decide what the 'correct' behavior is with regard to append-or-replace, and make the test ensure it does that.

**Comment by eddierubeiz** [comment]
I'm open to anything but I *do* want to avoid duplicate associations between collections and works. We may wish to modify the Kithe code to enforce that, too, but that would go in another PR.
Otherwise stated, you should not be able to create two identical rows in the `kithe_model_contains` table.

**Comment by jrochkind** [comment]
Yeah, I thought ActiveRecord used normally would prevent you from creating those duplicate join table rows, but I guess not? Anyway, that's a separate thing. But you already have a test ensuring that? Great, with the right tests, including for what I talked about above as far as delete-or-append, we can then refactor to be simpler while ensuring tests still pass.  (step three in TDD red -> green -> refactor!)

**Comment by eddierubeiz** [comment]
Oh, to answer your initial question:
_is this intending to replace any existing collection ids with the new one(s) specified, or append the new one to the existing ones?_
This is intended to append the new collection to the existing collections. But if a work in the cart _already_ belongs to the collection being appended, that work should not change, since... it's already in the collection.

**Comment by jrochkind** [comment]
> This is intended to append the new collection to the existing collections. But if a work in the cart already belongs to the collection being appended, that work should not change, since... it's already in the collection.

Great, is there a test that ensures that's what's happening, it's not overwriting existing collections? If so great! If not, can there be? Thanks!

**Comment by eddierubeiz** [comment]
Let me add that - good call.

File: spec/system/batch_edit_spec.rb [hunk 1]
```diff
@@ -58,20 +66,43 @@
...
+    expect(work1.contained_by.map(&:title)).to eq ["collection_1"]
 
     expect(work2.additional_title).to eq(["additional title 2a", "additional title 2b", "batch edit additional title"])
     expect(work2.provenance).to eq "batch edit provenance"
     expect(work2.creator).to eq([Work::Creator.new(category: "contributor", value: "creator 2")])
     expect(work2.description).to eq "description 2"
+    # Work 2 should have collection 1 added to it.
+    expect(work2.contained_by.first.title).to eq "collection_1"
+
+    # Go back to add collection 2 to both works:
```
Comment (jrochkind) [inline spec/system/batch_edit_spec.rb:92]: Adding another transaction in really slows down the test, makes it take longer.

Is there really anything we really need to do two batch edits (two submits) in order to test? 

Can't we just set up the initial data such that we can test everything we might want to test in only one batch edit operation? You have an initial condition of some records, you do a batch edit, test the post-edit condition to make sure it did all the things we wanted. Even adding more records if needed (although I'm not sure they are) will usually be faster/simpler than doing two batch edits. 

It's not clear to me what is being tested in this second batch edit -- didn't the first one already test adding a collection? If we want to test the ccondition of a record with a) no collections, b) a different collection, c) already had collection2 in it, can we just make three records with all three conditions for first batch edit?

Alternately, we could make a whole extra test case just for collection editing -- that's what "best practice" of test design would be... but it makes the test slower to do multiple requests like that. But packing multiple requests into just ONE test is the worst of both worlds...

Comment (eddierubeiz) [reply spec/system/batch_edit_spec.rb]: Fixed it!

**Pull request merged**